### PR TITLE
BUG: Add support to replace partitions in date-partitioned tables

### LINKED ISSUE #43: Support replacing partitions of date-partitioned tables
This issue is about adding support to write to partitions of BigQuery's [date-partitioned tables](https://cloud.google.com/bigquery/docs/partitioned-tables). Currently, reading is supported but inserts fails during execution of `GbqConnector.delete_and_recreate`: partition is deleted but creation fails due to existing table.

PR would include some refactoring to the current `to_gbq` function (e.g. getting rid of duplicate calls to BQ API).

Please see [mremes/pandas-gbq](https://github.com/mremes/pandas-gbq) for current changes. Currently not ready for a PR as I've encountered one issue with the BigQuery API (with subsequent create, delete, insertAll calls, insertAll's changes are not seen but after long waiting time).

### PR DESCRIPTION
**Pull request by ghost**
Closes #43

### CODE REVIEW

File: pandas_gbq/tests/test_gbq.py [hunk 1]
```diff
@@ -1093,6 +1096,79 @@ def test_upload_data_if_table_exists_replace(self):
                               project_id=_get_project_id(),
                               private_key=_get_private_key_path())
         assert result['num_rows'][0] == 5
+    
+    def test_upload_data_if_table_exists_replace_dpt_partition(self):
+        test_dpt_suffix = "20170101"
```
Comment (jreback) [inline pandas_gbq/tests/test_gbq.py]: can you add the issue number here as a comment

File: pandas_gbq/tests/test_gbq.py [hunk 2]
```diff
@@ -1093,6 +1096,79 @@ def test_upload_data_if_table_exists_replace(self):
...
+        dpt_partition = self.destination_date_partitioned_table + '$' + test_dpt_suffix
+
+        gbq.to_gbq(df, dpt_partition, _get_project_id(),
+                   chunksize=10000, private_key=_get_private_key_path())
+        
+        gbq.to_gbq(df_different_schema, dpt_partition,
+                   _get_project_id(), if_exists='replace',
+                   private_key=_get_private_key_path())
+        
+        sleep(30)
```
Comment (jreback) [inline pandas_gbq/tests/test_gbq.py]: would be nice to remove these things :> (maybe with a poll decorator?)

**Comment by jreback** [comment]
cc @parthea

File: pandas_gbq/tests/test_gbq.py [hunk 3]
```diff
@@ -1093,6 +1096,79 @@ def test_upload_data_if_table_exists_replace(self):
...
+
+        result1 = gbq.read_gbq("SELECT COUNT(*) AS num_rows FROM {0}"
+                               .format(dpt_partition),
+                               project_id=_get_project_id(),
+                               private_key=_get_private_key_path())
+
+        assert result1['num_rows'][0] == 15
+
+
+
```
Comment (parthea) [inline pandas_gbq/tests/test_gbq.py]: Please remove the extra blank lines

File: pandas_gbq/tests/test_gbq.py [hunk 4]
```diff
@@ -1117,7 +1193,7 @@ def test_google_upload_errors_should_raise_exception(self):
         with tm.assertRaises(gbq.StreamingInsertError):
             gbq.to_gbq(bad_df, self.destination_table + test_id,
                        _get_project_id(), private_key=_get_private_key_path())
-
+
```
Comment (parthea) [inline pandas_gbq/tests/test_gbq.py]: In the future please try to avoid spacing changes.

File: pandas_gbq/tests/test_gbq.py (hunk 2 above)
Comment (mremes) [inline pandas_gbq/tests/test_gbq.py]: Yup, would need to refactor some other tests also to get rid of `sleep` calls in between.

**Comment by jreback** [comment]
can you rebase

**Comment by mremes** [comment]
@jreback done

File: docs/source/changelog.rst [hunk 5]
```diff
@@ -8,6 +8,7 @@ Changelog
 - The dataframe passed to ```.to_gbq(...., if_exists='append')``` needs to contain only a subset of the fields in the BigQuery schema. (:issue:`24`)
 - Use the `google-auth <https://google-auth.readthedocs.io/en/latest/>`__ library for authentication because oauth2client is deprecated. (:issue:`39`)
 - ``read_gbq`` now has a ``auth_local_webserver`` boolean argument for controlling whether to use web server or console flow when getting user credentials. Replaces `--noauth_local_webserver` command line argument (:issue:`35`)
+- Add support to replace partitions in date-partitioned tables (:issue:`47`)
```
Comment (jreback) [inline docs/source/changelog.rst]: can you be slightly more verbose and include the link: [date-partitioned tables](https://cloud.google.com/bigquery/docs/partitioned-tables). any addition to the main docs about this? (or enhancements to the doc-string)?

**Comment by mremes** [comment]
@jreback it's green now. let's merge this before #46 as conflicts are solved.

**Comment by jreback** [comment]
can you fixup the linting failures?

**Comment by mremes** [comment]
@jreback done, coverage still lagging behind the target unfortunately...

File: docs/source/changelog.rst [hunk 6]
```diff
@@ -8,6 +8,7 @@ Changelog
 - The dataframe passed to ```.to_gbq(...., if_exists='append')``` needs to contain only a subset of the fields in the BigQuery schema. (:issue:`24`)
 - Use the `google-auth <https://google-auth.readthedocs.io/en/latest/>`__ library for authentication because oauth2client is deprecated. (:issue:`39`)
 - ``read_gbq`` now has a ``auth_local_webserver`` boolean argument for controlling whether to use web server or console flow when getting user credentials. Replaces `--noauth_local_webserver` command line argument (:issue:`35`)
+- Add support to replace partitions in `date-partitioned tables <https://cloud.google.com/bigquery/docs/partitioned-tables>`__. Partition must be specified with a partition decorator separator (``$``). (:issue:`47`)
```
Comment (jreback) [inline docs/source/changelog.rst]: can you add something in the docs about this (maybe the link above is enough) or a simple code-example.

File: pandas_gbq/gbq.py [hunk 7]
```diff
@@ -1032,17 +1032,30 @@ def to_gbq(dataframe, destination_table, project_id, chunksize=10000,
...
-        elif if_exists == 'append':
-            if not connector.schema_is_subset(dataset_id,
-                                              table_id,
-                                              table_schema):
-                raise InvalidSchema("Please verify that the structure and "
-                                    "data types in the DataFrame match the "
-                                    "schema of the destination table.")
+        else:
+            delay = 0
+            if not connector.verify_schema(dataset_id, table_id, table_schema):
```
Comment (parthea) [inline pandas_gbq/gbq.py]: This should be `if not connector.schema_is_subset(dataset_id, table_id, table_schema):`

See https://travis-ci.org/parthea/pandas-gbq/jobs/250715768:
The following unit test is failing :
```
pandas_gbq/tests/test_gbq.py::TestToGBQIntegrationWithServiceAccountKeyPath::test_upload_subset_columns_if_table_exists_append
```

**Comment by jreback** [comment]
@mremes if you can update.

**Comment by mremes** [comment]
@jreback sorry for the delay. I set up a dev environment on my machine, and I was able to successfully run the tests of the PR's feature. I've commited an updated branch but the build errors with `Invalid JWT Signature.` message. Is this something related to the service key used by CI?

**Comment by paoloburelli** [comment]
If the behaviour of 'append', 'replace' and 'fail' is referred to the existence of a partition I believe that this patch is incorrect.
The existence test is performed on the table rather than the partition, so writing to an empty partitioned table fails unless 'replace' or 'append' is specified.
Furthermore, the replace test is incorrect as it does not replace the content of a specific partition. 
If such a test was correctly performed it would fail as, due to a delay in the streaming buffer, deleting partition and streaming into it results in the data appearing after several minutes and easily up to an hour later.
I addressed all these issues in #85 with correct testing for all the writing policies.

**Comment by mremes** [comment]
@jreback @parthea This can be closed, #124 is for this now.

**Pull request closed without merging**